libraries/AP_Scripting/lua_bindings.cpp

```diff
@@ -532,3 +535,65 @@ int lua_removefile(lua_State *L) {
   const char *filename = luaL_checkstring(L, 1);
   return luaL_fileresult(L, remove(filename) == 0, filename);
 }
+
+
+// local cmd, data = crsf_pop() -- get byte array from crsf or nil if no data received
+int lua_crsf_pop(lua_State *L) {
+#if HAL_CRSF_TELEM_ENABLED
+    AP_RCProtocol_CRSF *crsf = AP_RCProtocol_CRSF::get_singleton();
+
+    // Allocate buffer on first call to this function
+    if (!crsf->lua_pop_buffer) {
+        crsf->lua_pop_buffer = new AP_RCProtocol_CRSF::lua_buffer_t();
+    }
+
+    // Move buffer data to lua
+    AP_RCProtocol_CRSF::lua_buffer_t *buf = crsf->lua_pop_buffer;
+    if (buf && buf->len > 0) {
+        WITH_SEMAPHORE(buf->sem);
+        lua_pushinteger(L, buf->cmd);
+        lua_pushlstring(L, buf->data, buf->len);
+        buf->len = 0;
+    } else {
+        lua_pushnil(L);
+        lua_pushnil(L);
+    }
+#else
+    lua_pushnil(L);
+    lua_pushnil(L);
+#endif
+    return 2;
+}
+
+// local num_bytes = crsfPush(cmd, data) -- push byte array to crsf
+int lua_crsf_push(lua_State *L) {
+#if HAL_CRSF_TELEM_ENABLED
+    int push_len = 0;
+    int isnum;
+    uint8_t cmd = lua_tointegerx(L, 1, &isnum);
+    size_t len;
+    const uint8_t *data = (uint8_t*)luaL_tolstring(L, 2, &len);
+    if(isnum && len>0) {
+        AP_CRSF_Telem *crsf = AP_CRSF_Telem::get_singleton();
+
+        // Allocate buffer on first call to this function
+        if(!crsf->lua_push_buffer) {
+            crsf->lua_push_buffer = new AP_RCProtocol_CRSF::lua_buffer_t();
+        }
+
+        // Move lua data to buffer if buffer is empty
+        AP_RCProtocol_CRSF::lua_buffer_t *buf = crsf->lua_push_buffer;
+        if (buf && buf->len == 0) {
+            WITH_SEMAPHORE(buf->sem);
+            push_len = (len < sizeof(AP_RCProtocol_CRSF::lua_buffer_t::data) ? len : sizeof(AP_RCProtocol_CRSF::lua_buffer_t::data));
```

Comment:
Better to error out if the length is too long rather than just throwing away the end of the message.

Reply:
Ah, I see, you return the number of bytes sent. But its a constant max lenght? Do we need that return?

Reply:
I agree, this can be done better. What do you think of this design?

data = crsf:pop()
data: nil=no data or alloc fail
(removed cmd return value because it is always 0x81 now)

result = crsf:push(data, [cmd=0x81])
result: true=ok, false=not pushed, nil=data too long or alloc fail
optional cmd argument

Reply:
I think alloc fail is OK to error out on (because I don't know what the script could do to fix), maybe data too long as well? If you try and send too much isn't that just a programming error? Then you just get a bool back?

Reply:
Alloc fail is a hard fault, but is it worth to waste flash for unlikely error messages? Or is there a pre-fabricated "out of memory" message available?

The data too long behaviour I (sort of) mirrored from EdgeTX:

```
int luaCrossfireTelemetryPush(lua_State * L)
...
  if (lua_gettop(L) > TELEMETRY_OUTPUT_BUFFER_SIZE ) {
    lua_pushboolean(L, false);
    return 1;
  }
```